### PR TITLE
Csharp iox2 12 ffi testing

### LINKED ISSUE #12: Testing of C# <-> Rust FFI
# Testing of C# <-> FFI Interface                                                                                               
                                                                                                                                  
## (Code) Example Of Cumbersome API                                                                                             
                                                                                                                                  
 The current FFI testing approach has several pain points:                                                                       
                                                                                                                                  
  1. **No systematic error path testing** - Tests only validate success paths, leaving FFI error codes and error translation      
  untested.                                                                                                                       
                                                                                                                                  
  2. **Union marshaling requires manual workarounds** - Complex union types like `iox2_static_config_t` cannot be properly        
  marshaled, requiring manual memory operations.                                                                                  
                                                                                                                                  
  3. **Handle lifecycle edge cases are untested** - No tests for double-disposal, use-after-free detection, or stress testing     
  handle creation/cleanup cycles.                                                                                                 
                                                                                                                                  
  4. **Callback testing is limited and not thread-safe** - Service discovery uses thread-local state (`_tempServiceList`) which is
   error-prone.                                                                                                                   
                                                                                                                                  
  ```csharp                                                                                                                       
  // Current approach: Only success paths tested, no FFI error validation                                                         
  [Fact]                                                                                                                          
  public void NativeLibraryLoads()                                                                                                
  {                                                                                                                               
      var exception = Record.Exception(() =>                                                                                      
      {                                                                                                                           
          Iox2NativeMethods.iox2_set_log_level_from_env_or(                                                                       
              Iox2NativeMethods.iox2_log_level_e.INFO);                                                                           
      });                                                                                                                         
      Assert.Null(exception);  // No error code validation                                                                        
  }                                                                                                                               
                                                                                                                                  
  // Current workaround for union marshaling limitations                                                                          
  // Note: Due to the simplified constructor (to avoid union marshaling issues),                                                  
  // PublishSubscribeConfig will be null. We can only verify the service list works.                                              
  // Pattern-specific configs will be null until we solve the union marshaling problem.                                           
                                                                                                                                  
  // Current handle passing pattern - unintuitive and error-prone                                                                 
  var publisherHandle = _handle.DangerousGetHandle();                                                                             
  var result = Native.Iox2NativeMethods.iox2_publisher_loan_slice_uninit(                                                         
      ref publisherHandle,  // Pass by reference for C to modify                                                                  
      IntPtr.Zero,                                                                                                                
      out var sampleHandle,                                                                                                       
      (UIntPtr)1);                                                                                                                
  // No validation of error codes beyond IOX2_OK  
```                                                                                
                                                                                                                                  
##  Improvement Suggestion                                                                                                          
                                                                                                                                  
  Introduce a comprehensive FFI testing framework that:                                                                           
                                                                                                                                  
  1. Validates all FFI error codes - Test each error path and ensure proper translation to C# error types.                        
  2. Tests handle lifecycle robustly - Include double-disposal detection, use-after-free scenarios, and concurrent handle         
  operations.                                                                                                                     
  3. Provides thread-safe callback testing utilities - Replace thread-local state with proper synchronization mechanisms.         
  4. Documents struct alignment validation - Add compile-time or runtime checks for struct size/alignment mismatches.             

```csharp                                                                                                                                  
  // Proposed: Comprehensive FFI error testing                                                                                    
  [Theory]                                                                                                                        
  [InlineData(iox2_service_builder_open_error_e.DOES_NOT_EXIST)]                                                                  
  [InlineData(iox2_service_builder_open_error_e.INSUFFICIENT_PERMISSIONS)]                                                        
  [InlineData(iox2_service_builder_open_error_e.INTERNAL_FAILURE)]                                                                
  public void FfiErrorCodes_AreProperlyTranslated(iox2_service_builder_open_error_e errorCode)                                    
  {                                                                                                                               
      var csharpError = Iox2ErrorTranslator.Translate(errorCode);                                                                 
      Assert.NotEqual(Iox2Error.Unknown, csharpError);                                                                            
      Assert.NotNull(csharpError.Message);                                                                                        
  }                                                                                                                               
                                                                                                                                  
  // Proposed: Handle lifecycle tests                                                                                             
  [Fact]                                                                                                                          
  public void SafeHandleDoubleDisposeDoesNotThrow()                                                                             
  {                                                                                                                               
      var handle = CreateTestHandle();                                                                                            
      handle.Dispose();                                                                                                           
      var exception = Record.Exception(() => handle.Dispose());                                                                   
      Assert.Null(exception);                                                                                                     
  }                                                                                                                               
                                                                                                                                  
  [Fact]                                                                                                                          
  public void SafeHandleUseAfterDisposeThrowsObjectDisposedException()                                                          
  {                                                                                                                               
      var publisher = CreateTestPublisher();                                                                                      
      publisher.Dispose();                                                                                                        
      Assert.Throws<ObjectDisposedException>(() => publisher.Loan<TestData>());                                                   
  }                                                                                                                               
                                                                                                                                  
  // Proposed: Thread-safe callback testing                                                                                       
  [Fact]                                                                                                                          
  public async Task ServiceDiscoveryConcurrentCallsAreThreadSafe()                                                              
  {                                                                                                                               
      var tasks = Enumerable.Range(0, 10)                                                                                         
          .Select(_ => Task.Run(() => ServiceDiscovery.List()))                                                                   
          .ToArray();                                                                                                             
                                                                                                                                  
      var results = await Task.WhenAll(tasks);                                                                                    
      Assert.All(results, r => Assert.NotNull(r));                                                                                
  }                                                                                                                               
                                                                                                                                  
  // Proposed: Struct alignment validation utility                                                                                
  [Fact]                                                                                                                          
  public void FfiStructsHaveCorrectSize()                                                                                        
  {                                                                                                                               
      // Validate C# struct sizes match C FFI expectations                                                                        
      Assert.Equal(                                                                                                               
          Iox2NativeMethods.iox2_node_builder_storage_t.Size,                                                                     
          FfiStructValidator.GetExpectedSize("iox2_node_builder_storage_t"));                                                     
  } 
```

### PR DESCRIPTION
**Pull request by patdhlk**
## Notes for Reviewer

- **Core change:** introduces `Iceoryx2.Native.CallbackContext` (GCHandle-pinned token) and routes three FFI callback sites (`Node.List`, `WaitSet.WaitAndProcessOnce`, `WaitSet.WaitAndProcessOnce(timeout)`, `WaitSet.WaitAndProcessInternal`) through it.
- **Bugs fixed:**
  - `Node.List()` used a static `_tempServiceList` field — two concurrent callers on the same process could share a list. Regression test: `ServiceDiscoveryConcurrentTests`.
  - `WaitSet._nativeCallback` was a single instance field rewritten on every `WaitAndProcessOnce*`/`WaitAndProcessInternal` call — a second overlapping call would overwrite the first's closure. Now replaced by two `static readonly` trampolines + per-call pinned context.
  - `IOX2_SERVICE_ID_LENGTH` was 32 but the cbindgen header defines it as 64 (`iceoryx2.h` line 69). `Node.List()`'s manual marshal was reading the `name` field at the wrong offset. No existing test caught this because `ServiceDiscoveryTests` deliberately avoided the `name` field.
- **Focus points when reviewing:**
  - `CallbackContext.Pin/Unpin` pairing in every caller (`Pin` outside `try`, `Unpin` in `finally`).
  - Trampoline `catch { return STOP; }` — intentionally silent; swallowing user-callback exceptions is required at the FFI boundary. Pinned in tests (`WaitSet_TrampolineReturnsStop_WhenCallbackThrows`).
  - `NativeStructSizesTests` hardcodes 9104 / 18696 with header references — treat as a tripwire, update in lockstep with cbindgen output.

## Pre-Review Checklist for the PR Author

* [x] Add sensible notes for the reviewer
* [x] PR title is short, expressive and meaningful
* [x] Consider switching the PR to a draft (`Convert to draft`)
    * as draft PR, the CI will be skipped for pushes
* [x] Relevant issues are linked in the [References](#references) section
* [x] Branch follows the naming format (`csharp-iox2-123-short-description`)
* [x] Commits messages are according to this [guideline][commit-guidelines]
    * [x] Commit messages have the issue ID (`[#123] Add feature description`)
* [x] Tests have been added/updated for new functionality
* [x] XML documentation comments added to public APIs
* [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
* [x] Assign PR to reviewer
* [x] All checks have passed

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2-csharp/blob/main/CHANGELOG.md

## PR Reviewer Reminders

* Commits are properly organized and messages are according to the guideline
* Unit tests have been written for new behavior
* Public API is documented with XML comments
* PR title describes the changes
* Code follows C# conventions (PascalCase for public APIs)
    * `dotnet format` has been exectued before submitting

## References

Closes #12

### CODE REVIEW

File: src/Iceoryx2/Native/CallbackContext.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2025 Contributors to the Eclipse Foundation
```
Comment (dkroenke) [inline src/Iceoryx2/Native/CallbackContext.cs]: Should be 2026, same for the other newly created source files

File: CHANGELOG.md
Comment (dkroenke) [inline CHANGELOG.md:24]: Does this really fits to the issue description? For a better readability, it would be better to split the text into separate bullet points.

**Comment by patdhlk** [comment]
@dkroenke do you have time to review this one?

File: src/Iceoryx2/Node.cs [hunk 2]
```diff
@@ -51,49 +51,38 @@ public sealed class Node : IDisposable
...
-            if (configPtr == IntPtr.Zero || _tempServiceList == null)
+            var list = CallbackContext.Peek<List<ServiceStaticConfig>>(context);
+            if (list is null || configPtr == IntPtr.Zero)
             {
                 return Iox2NativeMethods.iox2_callback_progression_e.STOP;
             }
 
-            // Read the basic fields directly from memory without creating the full struct
-            // to avoid issues with the union in iox2_static_config_t
-
```
Comment (dkroenke) [inline src/Iceoryx2/Node.cs:73]: Why are these comments removed?

File: src/Iceoryx2/Native/Iox2NativeMethods.cs [hunk 3]
```diff
@@ -68,7 +68,7 @@ private static IntPtr DllImportResolver(string libraryName, Assembly assembly, D
     internal const int IOX2_OK = 0;
     internal const int IOX2_NODE_NAME_LENGTH = 128;
     internal const int IOX2_SERVICE_NAME_LENGTH = 255;
-    internal const int IOX2_SERVICE_ID_LENGTH = 32;
+    internal const int IOX2_SERVICE_ID_LENGTH = 64;
```
Comment (dkroenke) [inline src/Iceoryx2/Native/Iox2NativeMethods.cs:71]: Just a general question, is there a possibility to read or import the C header to take the values directly from the source? That could possibly reduce the maintenance burden instead of duplicating the values.

File: src/Iceoryx2/Node.cs [hunk 4]
```diff
@@ -51,49 +51,38 @@ public sealed class Node : IDisposable
...
-            var serviceConfig = new ServiceStaticConfig(id, name, messagingPattern);
-            _tempServiceList.Add(serviceConfig);
+            list.Add(new ServiceStaticConfig(id, name, messagingPattern));
             return Iox2NativeMethods.iox2_callback_progression_e.CONTINUE;
         }
-        catch (Exception ex)
+        catch
         {
-            Console.WriteLine($"Error in service list callback: {ex.Message}");
-            Console.WriteLine($"Stack trace: {ex.StackTrace}");
```
Comment (dkroenke) [inline src/Iceoryx2/Node.cs:96]: Wouldn't it help for debugging to have a stack trace?

File: tests/Ffi/NativeStructSizesTests.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+///
+/// Source of truth: iceoryx2/target/release/iceoryx2-ffi-c-cbindgen/include/iox2/iceoryx2.h
+/// Each constant below mirrors a `uint8_t internal[N]` storage-struct size in
+/// that header. If the Rust FFI changes a storage size, the header regenerates,
+/// the bindings must be updated, and both the Size attribute in
+/// Iox2NativeMethods.cs and the constant below must be updated together.
+/// </summary>
+public class NativeStructSizesTests
+{
+    private const int ServiceBuilderStorageSize = 9104;   // iceoryx2.h iox2_service_builder_storage_t
```
Comment (dkroenke) [inline tests/Ffi/NativeStructSizesTests.cs:36]: Is it possible to share the magic numbers from `src/Iceoryx2/Native/Iox2NativeMethods.cs` with this file somehow? That would make it less error-prone in case the values are changing.

File: src/Iceoryx2/Native/CallbackContext.cs (hunk 1 above)
Comment (patdhlk) [reply src/Iceoryx2/Native/CallbackContext.cs]: Done — bumped to 2026 across the six newly added files (CallbackContext.cs, FfiAssert.cs, NativeStructSizesTests.cs, WaitSetConcurrentTests.cs, CallbackContextTests.cs, ServiceDiscoveryConcurrentTests.cs) in 7bdd9fd.

File: CHANGELOG.md
Comment (patdhlk) [reply CHANGELOG.md:24]: Split into three per-fix bullets in 3f2d914.

On the issue-fit question: #12 is about FFI testing in general and explicitly calls out two of the bugs ("Callback testing is limited and not thread-safe — Service discovery uses thread-local state (`_tempServiceList`) which is error-prone" → covered by the Node.List + WaitSet fixes; and "Union marshaling requires manual workarounds" → the SERVICE_ID_LENGTH offset fix is the same code path), so the bugfixes are in scope for the issue. Happy to move them under a separate follow-up issue if you'd prefer cleaner accounting.

File: src/Iceoryx2/Node.cs (hunk 2 above)
Comment (patdhlk) [reply src/Iceoryx2/Node.cs:73]: Good catch — the offset comments did real work explaining the manual marshal. Restored in 708cc02.

File: src/Iceoryx2/Node.cs (hunk 4 above)
Comment (patdhlk) [reply src/Iceoryx2/Node.cs:96]: Good point on the diff between callback layers — the trampolines in `WaitSet` intentionally swallow everything because they wrap arbitrary user code at the FFI boundary (a user-thrown `NullReferenceException` must not unwind into native iceoryx2). But `ServiceListCallback` only runs our own marshalling code; if it throws, that is a binding bug and silencing the stack trace makes it un-debuggable.

Restored the diagnostic logging in 493c2de, this time on `Console.Error` so it doesn't pollute the stdout-driven console logger contract. Will return `STOP` on failure as before.

File: src/Iceoryx2/Native/Iox2NativeMethods.cs (hunk 3 above)
Comment (patdhlk) [reply src/Iceoryx2/Native/Iox2NativeMethods.cs:71]: Fair point — and that is exactly the same maintenance pain that motivated `NativeStructSizesTests` (the storage-size tripwire). A few directions, ranked by realism:

1. **ClangSharp / CppSharp source-gen.** Run cbindgen → C header → ClangSharp at build time and emit `Iox2NativeMethods.Generated.cs`. This is the proper fix but adds an LLVM/libclang build dependency and a generator project. Probably needs its own issue.
2. **A `cbindgen --emit-cmake-config`-style sidecar.** Have the FFI build also emit a small `.json` of constants + struct sizes that a tiny C# T4 / source generator consumes. Lower-weight than ClangSharp but still needs a tool we don't yet have.
3. **Status quo + tripwires.** What's in this PR: the manual `const` mirror plus `NativeStructSizesTests` to catch drift. Cheap, but only catches what we explicitly bookkeep.

I'd lean toward (1) as a follow-up but don't like it at the moment — happy to open an issue for it. For this PR I think the tripwire approach is the right scope, but let me know if you'd rather hold this one until the generator lands.

File: tests/Ffi/NativeStructSizesTests.cs (hunk 5 above)
Comment (patdhlk) [reply tests/Ffi/NativeStructSizesTests.cs:36]: I'd push back on this one — sharing erases the tripwire that the test is supposed to be. `Marshal.SizeOf<T>` reads the `StructLayout(Size = ...)` attribute, so if both the attribute and the test reference the same shared const, the assert collapses to `const == const` and trivially passes.

The double-entry duplication is load-bearing: the failure mode it catches is exactly "someone updated the C# binding's `Size` to track new cbindgen output but forgot the test constant (or vice versa)". Killing the duplication kills the only thing forcing the maintainer to touch both files.

In the long run the right fix is generating both sides from the C header (see the thread on `IOX2_SERVICE_ID_LENGTH`). Until then I'd keep the duplication, but I added an inline note at the literals (eebaffe) so the next reader doesn't have the same "why aren't these shared?" reaction. Let me know if you'd rather I go the other way.

File: CHANGELOG.md
Comment (dkroenke) [reply CHANGELOG.md:24]: The list of bugfixes is good but the main reason for this Pull-Request is missing.

File: src/Iceoryx2/Native/Iox2NativeMethods.cs (hunk 3 above)
Comment (dkroenke) [reply src/Iceoryx2/Native/Iox2NativeMethods.cs:71]: Would you create an issue out of it?

File: src/Iceoryx2/Native/CallbackContext.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        if (token == IntPtr.Zero)
+            return null;
+        var handle = GCHandle.FromIntPtr(token);
+        return handle.IsAllocated ? handle.Target as T : null;
+    }
+
+    /// <summary>
+    /// Frees the <see cref="GCHandle"/> and returns the (now unrooted) state.
+    /// MUST be called exactly once per <see cref="Pin{T}"/>, in a <c>finally</c>.
```
Comment (dkroenke) [inline src/Iceoryx2/Native/CallbackContext.cs:74]: There could be a test case for it

File: src/Iceoryx2/Native/CallbackContext.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    /// </remarks>
+    public static T? Unpin<T>(IntPtr token) where T : class
+    {
+        if (token == IntPtr.Zero)
+            return null;
+        var handle = GCHandle.FromIntPtr(token);
+        if (!handle.IsAllocated)
+            return null;
+        var target = handle.Target as T;
+        handle.Free();
```
Comment (dkroenke) [inline src/Iceoryx2/Native/CallbackContext.cs:96]: There could be test cases against the preliminary checks to make sure the API will not be misused.
Do we have here similar undefined behavior compared to C/C++ when `handle.Free()` would be called two times on the same handle?

File: tests/Ffi/FfiAssert.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    /// and includes a caller-supplied decoded error name on failure.
+    /// </summary>
+    public static void Ok(int result, string operation, Func<int, string> decode)
+    {
+        if (result != Iox2NativeMethods.IOX2_OK)
+        {
+            throw new XunitException(
+                $"FFI call '{operation}' failed: code={result} ({decode(result)})");
+        }
+    }
```
Comment (dkroenke) [inline tests/Ffi/FfiAssert.cs:48]: Are these functions added as preparation to be later used somewhere?

File: CHANGELOG.md
Comment (patdhlk) [reply CHANGELOG.md:24]: The PR description and the linked issue (#12) cover the test-infrastructure intent (CallbackContext + FfiAssert + NativeStructSizesTests + the WaitSet/ServiceDiscovery concurrent tests), so I'd prefer to keep the CHANGELOG focused on user-visible changes — happy to add a Features entry if you'd rather it be redundant there too.

File: src/Iceoryx2/Native/Iox2NativeMethods.cs (hunk 3 above)
Comment (patdhlk) [reply src/Iceoryx2/Native/Iox2NativeMethods.cs:71]: Opened #21 with the three options (ClangSharp source-gen / cbindgen JSON sidecar / status quo + tripwires) plus a recommendation.

File: src/Iceoryx2/Native/CallbackContext.cs (hunk 6 above)
Comment (patdhlk) [reply src/Iceoryx2/Native/CallbackContext.cs:74]: Added in b2a8498 — `CallbackContextTests` now also covers `Pin(null)`, `Peek<WrongType>`, and `Unpin<WrongType>` (the last one with a WeakReference tripwire to confirm the GCHandle is still Freed even on type mismatch). The existing 10k-iteration leak test stays as the round-trip happy-path check.

File: src/Iceoryx2/Native/CallbackContext.cs (hunk 7 above)
Comment (patdhlk) [reply src/Iceoryx2/Native/CallbackContext.cs:96]: Good question — I went and probed it before answering, and the result is worse than I expected: yes, this has the same shape as a C double-free.

After `handle.Free()`, re-deriving from the same `IntPtr` via `GCHandle.FromIntPtr` returns a struct whose `IsAllocated` reports **true**. The runtime aggressively recycles handle slots, so the token can resolve to whatever pin now owns that slot. Calling `Free` on the stale token then silently frees the new owner's pin — classic use-after-free on a recycled GC handle index. The `if (!handle.IsAllocated)` preliminary check in `Unpin` effectively only defends against `IntPtr.Zero` in practice; it cannot detect a recycled slot.

Probe (Release, .NET 9):

```
After Free, FromIntPtr.IsAllocated=True
Target=EventPipeEventProvider     // an unrelated runtime object now occupying the slot

After re-alloc, t2a==t2b: True    // re-Alloc gives us the same slot back
Stale.IsAllocated=True
Stale.Target?.Length=1            // sees the *new* pin's content
Stale.Free() succeeded            // silently frees the new pin
```

So the proper enforcement is discipline — the documented `try`/`finally` pairing — not a runtime check. I beefed up the XML doc on `CallbackContext` to say so explicitly ("treat the token like a C pointer; double-Unpin and Peek-after-Unpin are the .NET analogue of use-after-free / double-free") and added the tests we *can* enforce in b2a8498:

  - `Pin(null)` → throws `ArgumentNullException`
  - `Peek<WrongType>(token)` → returns null, original pin still intact for the correct `T`
  - `Unpin<WrongType>(token)` → returns null but the `GCHandle` is still `Free`d (WeakReference tripwire confirms no leak on type mismatch)

If you want stronger active defense (e.g. a `ConcurrentDictionary<IntPtr, ...>` tracking outstanding tokens), I'd open a follow-up since it'd change the perf profile and ownership model — happy to file that if you think we need it on top of the discipline contract.

File: tests/Ffi/FfiAssert.cs (hunk 8 above)
Comment (patdhlk) [reply tests/Ffi/FfiAssert.cs:48]: Yes, but only loosely — they were sitting unused because the new concurrent tests under `tests/Ffi/` go through `Iox2.Run().IsOk` rather than raw FFI codes. Wired `FfiAssert.Ok` into the two raw-FFI sites in `RuntimeTests.cs` in 6ded0f6 so the helper has concrete callers and failures carry the operation name; future raw-FFI tests can pick it up too.

**Review by dkroenke — APPROVED** [review]
